src/server/gc_worker/gc_worker.rs

```diff
@@ -422,6 +454,139 @@ where
+                    if key < region.get_start_key() {
+                        self.keys.next();
+                    } else if region.get_end_key().is_empty() || key < region.get_end_key() {
+                        return self.keys.next();
+                    } else {
+                        self.regions.next();
+                    }
+                }
+            }
+        }
+
+        fn get_keys_in_regions(
+            keys: Vec<Key>,
+            regions_provider: Option<(u64, Arc<dyn RegionInfoProvider>)>,
+        ) -> Result<Box<dyn Iterator<Item = Key>>> {
+            if keys.len() >= 2 {
+                if let Some((store_id, region_info_provider)) = regions_provider {
+                    let start = keys.first().unwrap().as_encoded();
+                    let end = keys.last().unwrap().as_encoded();
+                    let regions = box_try!(region_info_provider.get_regions_in_range(start, end))
+                        .into_iter()
+                        .filter(move |r| find_peer(r, store_id).is_some())
+                        .peekable();
+
+                    let keys = keys.into_iter().peekable();
+                    return Ok(Box::new(KeysInRegions { keys, regions }));
+                }
+            }
+            Ok(Box::new(keys.into_iter()))
+        }
+
+        let mut keys = get_keys_in_regions(keys, regions_provider)?;
+        //--------------------------------KeysInRegions-----------------------------
+
+        let mut snapshot = self
+            .engine
+            .snapshot_on_kv_engine(range_start_key, range_end_key)?;
+
+        let mut raw_modifys = MvccRaw::new();
+
+        let mut next_gc_key = keys.next();
+        while let Some(ref key) = next_gc_key {
+            if let Err(e) = self.raw_gc_key(safe_point, key, &mut raw_modifys, &mut snapshot) {
+                GC_KEY_FAILURES.inc();
+                error!(?e; "Raw GC meets failure"; "key" => %key,);
+                // Switch to the next key if meets failure.
+            }
+            // flush_raw_gc;
+            Self::flush_raw_gc(raw_modifys, &self.limiter, &self.engine)?;
+            // after flush, reset rwModify=let RawModify=MvccRaw::new();
+            raw_modifys = MvccRaw::new();
+            next_gc_key = keys.next();
+        }
+        Ok((0, 0))
+    }
+
+    fn raw_gc_key(
+        &mut self,
+        safe_point: TimeStamp,
+        key: &Key,
+        raw_modifys: &mut MvccRaw,
+        kv_snapshot: &mut <E as Engine>::Snap,
+    ) -> Result<()> {
+        let start_key = key
+            .clone()
+            .append_ts(TimeStamp::new(safe_point.into_inner() - 1));
+
+        let end_key = key.clone().append_ts(TimeStamp::zero());
+
+        let start_key_arr = start_key.as_encoded();
+        let start_seek_key = Key::from_encoded_slice(start_key_arr);
+
+        let mut cursor = CursorBuilder::new(kv_snapshot, CF_DEFAULT)
+            .range(Some(start_key), Some(end_key))
```

Comment:
Emmm, I'm not sure if keys with timestamp appended can be used as the range of the cursor safely.